numpy/ma/core.py

```diff
@@ -3189,23 +3189,50 @@ def __getitem__(self, indx):
         # mask of being reshaped if it hasn't been set up properly yet
         # So it's easier to stick to the current version
         _mask = self._mask
+
+
+        # self contains an object array of arrays (yes, that happens).
+        # unfortunately, this fails if the inner arrays themselves are also
+        # object arrays.
+        is_array_elem_of_object_array = (
+            self.dtype.type is np.object_ and
+            isinstance(dout, np.ndarray) and
+            self.dtype is not dout.dtype
+        )
+
+        if _mask is not nomask:
+            # _mask cannot be a subclass, so it tells us whether we should
+            # expect a scalar. It also cannot be of dtype object.
+            mout = self._mask[indx]
+            expect_scalar = not isinstance(mout, np.ndarray)
+        else:
+            # much harder to tell if we should get a scalar here - not quite
+            # correct
```

Comment:
In an ideal world, we would call `getmaskarray()`, and do the same as the above code path. Unfortunately, this would add a hefty performance cost for the most common cases.